staging/src/k8s.io/client-go/util/homedir/homedir.go

```diff
@@ -21,27 +21,49 @@ import (
 	"runtime"
 )
 
-// HomeDir returns the home directory for the current user
+// HomeDir returns the home directory for the current user.
 func HomeDir() string {
+	return PreferredHomeDir(nil)
+}
+
+// PreferredHomeDir returns the home directory for the current user.
+// If there are multiple possible home directories (as on windows), all possible paths
+// are tested with each path test in order, and the first path to pass a test is returned.
+// If no paths pass any tests, or no path tests are provided, the first possible path is returned.
+func PreferredHomeDir(pathTests []func(string) bool) string {
 	if runtime.GOOS == "windows" {
 
+		possiblePaths := []string{}
 		// First prefer the HOME environmental variable
 		if home := os.Getenv("HOME"); len(home) > 0 {
 			if _, err := os.Stat(home); err == nil {
-				return home
+				possiblePaths = append(possiblePaths, home)
 			}
 		}
 		if homeDrive, homePath := os.Getenv("HOMEDRIVE"), os.Getenv("HOMEPATH"); len(homeDrive) > 0 && len(homePath) > 0 {
 			homeDir := homeDrive + homePath
 			if _, err := os.Stat(homeDir); err == nil {
-				return homeDir
+				possiblePaths = append(possiblePaths, homeDir)
 			}
 		}
 		if userProfile := os.Getenv("USERPROFILE"); len(userProfile) > 0 {
```

Comment:
I believe USERPROFILE should be the second test in this list. HOME is never present by default and is set by the user. USERPROFILE is always present (in mordern Windows versions).

Reply:
reordering breaks backward compatibility with currently working configurations. an attempt was made to do that in https://github.com/kubernetes/kubernetes/pull/67403, but is a behavioral change I think we should avoid.

Reply:
LGTM, as I don't have enough data to disagree. Having the writable test on HOMEDRIVE/HOMEPATH will do the job for us, but this combination seems a relic from an old past. By curiosity, how many users would be broken by this ?

Reply:
I believe that the only thing that will fix #61946 is purely order. The problem is that the order should match that of other clients that interact with the `.kube` folder. If `HOMEDRIVE/HOMEPATH` is writable it should still be ignored if it does not have a `.kube` folder. Documentation (while dated, and sparse) on why was provided here: https://github.com/kubernetes/kubernetes/issues/61946#issuecomment-412956711

Reply:
> If `HOMEDRIVE/HOMEPATH` is writable it should still be ignored if it does not have a `.kube` folder

i think this is the key for backwards compat here.

- check HOME
- check if HOMEDRIVE + HOMEPATH exists
  - write a flag (e.g. isHomeDriveHomePathValid) that this path exists (i.e ENV variables are sane)
  - if .kube exists there respect that (backwards compat)
  - if not falltrough
- check USERPROFILE
- if isHomeDriveHomePathValid == true, use that
- return HOME (backwards compat)

would that work?

also is it possible to bump client-go with `action required` type of a release note in case of a breaking change?

Reply:
I believe this would work. There is definitely a risk here in some corporate environments. If it is documented that users would need to manually move / remove a `.kube` folder then I think that works. It puts the burden on the users but keeps working as previous versions.

Reply:
here are the versioning rules for client-go:
https://github.com/kubernetes/client-go#versioning

so what we are trying to do is a bug fix, yet it's under question how much we are breaking.
existing users can continue using HOMEDRIVE+PATH, while new users can land in USERPROFILE.

also looking at the versioning table and as far as i understand a MAJOR bump will happen with k8s 1.14. so if we consider this a breaking change it can be safely added to this MAJOR.

Reply:
It might be ok to break client-go, but not kubectl, which also depends on this behavior.  kubectl backwards compatibility is a pretty strong guarantee, and breaking existing users especially scripting people may have developed has significant implications.

For instance, if someone has a script that cleans up namespaces on a dev system that uses the old config, and we update kubectl, and their new config points to production, and suddenly the script is running against production.  While this may seem unlikely, we've already had breaks like this in the last few years in various areas.

Reply:
that's true. i forgot about kubectl's guarantees!

Reply:
So is the thought that if the logic implemented a check to continue using `HOMEDRIVE/HOMEPATH` where users have that, and a `.kube` folder exists, that this would break that usage? This should preserve the current functionality, and just forces new implementations to be in `USERPROFILE`?

Would it make sense to check all possible homePaths for a `.kube` folder and if multiple paths have a `.kube` folder, then exist with a detailed error?

Reply:
> Would it make sense to check all possible homePaths for a .kube folder and if multiple paths have a .kube folder, then exist with a detailed error?

That is also breaking. A general principle is that configurations that work today should continue to work.

Reply:
Is there a reason the check implemented by this PR doesn't work? It allows existing configurations to continue working, prefers USERPROFILE if it contains a kubeconfig file and HOMEDRIVE/HOMEFOLDER doesn't, and if no locations contain a kubeconfig, prefers a writeable home folder. 

What scenario is unaddressed by that?

Reply:
The only issue I see is that it will still prefer `HOMEDRIVE/HOMEPATH` if it is writable. The biggest issue here is that other tools and languages that modify the `.kube` (cloud provider auth tools) do not follow this preference.

To deprecate `HOMEDRIVE/HOMEPATH` but still support it, I think it needs to prefer `USERPROFILE` when both are writable and no `.kube` exists.

Reply:
ok, took a stab at that:

1. the first of %HOME%, %HOMEDRIVE%%HOMEPATH%, %USERPROFILE% containing a `.kube\config` file is returned.
2. if none of those locations contain a `.kube\config` file, the first of %HOME%, %USERPROFILE%, %HOMEDRIVE%%HOMEPATH% that exists and is writeable is returned.
3. if none of those locations are writeable, the first of %HOME%, %USERPROFILE%, %HOMEDRIVE%%HOMEPATH% that exists is returned.
4. if none of those locations exists, the first of %HOME%, %USERPROFILE%, %HOMEDRIVE%%HOMEPATH% that is set is returned.

Reply:
if that looks write, I'll add unit tests

Reply:
That sounds correct to me. There are probably some edge cases that could exist from the comments above, but I think that would provide suitable protection for backwards compatibility in environments that use alternate home paths.

Reply:
LGTM